src/main/scala/mesosphere/marathon/api/forwarder/AsyncUrlConnectionRequestForwarder.scala

```diff
@@ -0,0 +1,210 @@
+package mesosphere.marathon
+package api.forwarder
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.{ContentType, ContentTypes, HttpEntity, HttpResponse}
+import akka.http.scaladsl.settings.{ClientConnectionSettings, ConnectionPoolSettings}
+import akka.http.scaladsl.{Http, HttpsConnectionContext}
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{HttpHeader, HttpMethod, HttpMethods, HttpRequest, Uri}
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.Source
+import com.typesafe.sslconfig.akka.AkkaSSLConfig
+import java.net._
+import javax.net.ssl._
+import javax.servlet.AsyncContext
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+import akka.Done
+import akka.http.scaladsl.model.StatusCodes._
+import com.google.common.io.Closeables
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.api.LeaderProxyFilter
+import mesosphere.marathon.stream.Implicits._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+/**
+  * Forwarder which uses Akka HTTP to proxy requests, and then
+  */
+class AsyncUrlConnectionRequestForwarder(
+    sslContext: SSLContext,
+    leaderProxyConf: LeaderProxyConf,
+    myHostPort: String)(implicit executionContext: ExecutionContext, actorSystem: ActorSystem)
+  extends RequestForwarder with StrictLogging {
+  private implicit val mat = ActorMaterializer()
+
+  import RequestForwarder._
+
+  private[this] val viaValue: String = s"1.1 $myHostPort"
+
+  private val sslConfig = AkkaSSLConfig().mapSettings { s =>
+    if (leaderProxyConf.leaderProxySSLIgnoreHostname())
+      s.withHostnameVerifierClass(classOf[IgnoringHostnameVerifier])
+    else
+      s
+  }
+  private val connectionContext = new HttpsConnectionContext(sslContext, sslConfig = Some(sslConfig))
+  private val connectionSettings = ClientConnectionSettings(actorSystem).
+    withIdleTimeout(leaderProxyConf.leaderProxyReadTimeout().millis).
+    withConnectingTimeout(leaderProxyConf.leaderProxyConnectionTimeout().millis)
+  private val poolSettings = ConnectionPoolSettings(actorSystem)
+    .withConnectionSettings(connectionSettings)
+    .withMaxRetries(0)
+
+  class IgnoringHostnameVerifier extends javax.net.ssl.HostnameVerifier {
```

Comment:
What is this doing?

Reply:
If you search for the reference code (there is only one), this is the hostname verifier used if the ignore hostname authenticity Marathon configuration option is enabled